### PR TITLE
Count metric for completed jobs

### PR DESCRIPTION
**Pull request by MadsNielsen**
Here you go @JohnLZeller . This is the pull request for the 'count' metric.

The PR also includes the changes for the tags...which this requires. The relevant part to review would then be in the second commit of this PR.

### CODE REVIEW

**Comment by JohnLZeller** [comment]
Hey @MadsNielsen!

Okay, so this PR works, but it is not the ideal way of doing this. First of all, it's not entirely accurate to use a gauge in such a way, and it'll require using rollups and things on a dashboard to make it work.

Additionally, multiple jobs finishing at the same second will not be counted correctly.
This works fine as a forked hack if someone wants to utilize it, but I have an alternative suggestion.

Let's shoot out a UDP packet to the local Dogstatsd (packaged with the datadog agent). If it's not there, the UDP packets will fall on nothing, if it is there, then the counts will be pushed up to Datadog.

Let me know if I can clarify anything here for you.

**Comment by MadsNielsen** [comment]
@JohnLZeller Sounds fine by me. I assume this is documented in your API docs so i can implement it?

**Comment by JohnLZeller** [comment]
@MadsNielsen It's not in the API docs actually, because it's an agent feature, not a feature of the API.
http://docs.datadoghq.com/guides/dogstatsd/

**Comment by MadsNielsen** [comment]
Hey @JohnLZeller 

I've started working on this, and a couple of questions popped up. 

1. We probably want an url // port configuration for the statsdaemon jenkins should communicate with. We can the just default to localhost // 8125

2. What about tags, i assume we need to attach all the same tags as we do to the build finished. Atleast job name so you can group by job name with this counter.

But i think i got the overall idea on how this works so i'm working on getting this done :)

**Comment by MadsNielsen** [comment]
Hey @JohnLZeller 

Another question popped up: What is your thought of using one of the excellent client libraries you're linking to in the plugin? 

More specifically im lookin at this: https://github.com/indeedeng/java-dogstatsd-client

**Comment by JohnLZeller** [comment]
@MadsNielsen Yeah, but let's use our own client fork, https://github.com/DataDog/java-dogstatsd-client

1. Yeah let's add this as an optional global configuration, with those defaults. Let's hide this behind an Advanced options dropdown.

2. Yeah, let's include our tags.

**Comment by JohnLZeller** [comment]
@MadsNielsen once you've implemented the new functionality we spoke of, would you rebase this PR from master (I merged your other changes)? Thanks :)

**Comment by MadsNielsen** [comment]
@JohnLZeller Okay. First draft for this feature is ready. I used your client fork for this part. The configuration now includes an `Advanced` section where you can specify the address to your daemon. 

I found the Client reasonably ok to work with, one issue i had initially was that i just couldn't just create a new client in the `onCompleted` method. Since a run listener is global (one instance per Jenkins) i found that if i created the client there, and stopped is at well...it would cancel out the udp requests... :)

So...i think i figured out how to do it properly. When the global configuration page is saved. We need to pick up eventual configuration changes and create a new client for our run listener to use. And when we use the Client in the run listener, we can ask the descriptor to give us the current configured instance...

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java [hunk 1]
```diff
@@ -447,6 +529,22 @@ public boolean configure(final StaplerRequest req, final JSONObject formData)
...
+      //The global configuration
+      if (client != null) {
+        try {
+          client.stop();
+          String hp = daemonHost.split(":")[0];
+          int pp = Integer.parseInt(daemonHost.split(":")[1]);
+          client = new NonBlockingStatsDClient("jenkins.job", hp, pp);
+          logger.finer(String.format("Created new dogstatsdaemon client (%s:%S)!", hp, pp));
+        } catch (Exception e) {
+          logger.log(Level.SEVERE, "Unable to crete new DogstatsClient", e);
```
Comment (JohnLZeller) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java]: create*

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java [hunk 2]
```diff
@@ -447,6 +529,22 @@ public boolean configure(final StaplerRequest req, final JSONObject formData)
         tagNode = false;
       }
 
+      daemonHost = formData.getString("daemonHost");
+      //When form is saved...reinitialize the StatsDClient.
+      //We need to stop the old one first. And crete a new one with the new data from
```
Comment (JohnLZeller) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java]: create*

File: src/main/resources/org/datadog/jenkins/plugins/datadog/DatadogBuildListener/global.jelly [hunk 3]
```diff
@@ -30,5 +30,10 @@
              description="Name of node the current build is running on (i.e. 'master' for master node)." >
       <f:checkbox title="node" field="tagNode" optional="true" />
     </f:entry>
+    <f:advanced>
+        <f:entry title="Dogstats daemon host" field="daemonHost">
```
Comment (JohnLZeller) [inline src/main/resources/org/datadog/jenkins/plugins/datadog/DatadogBuildListener/global.jelly]: Title should instead be `DogStatsD Host`

File: src/main/resources/org/datadog/jenkins/plugins/datadog/DatadogBuildListener/help-daemonHost.html [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Enter your Datadog statsdaemon host specification. In the form <em>hostname:port</em>
```
Comment (JohnLZeller) [inline src/main/resources/org/datadog/jenkins/plugins/datadog/DatadogBuildListener/help-daemonHost.html]: `Enter your DogStatsD host specification, in the form <em>hostname:port</em>`

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java [hunk 5]
```diff
@@ -447,6 +529,22 @@ public boolean configure(final StaplerRequest req, final JSONObject formData)
...
+      //The global configuration
+      if (client != null) {
+        try {
+          client.stop();
+          String hp = daemonHost.split(":")[0];
+          int pp = Integer.parseInt(daemonHost.split(":")[1]);
+          client = new NonBlockingStatsDClient("jenkins.job", hp, pp);
+          logger.finer(String.format("Created new dogstatsdaemon client (%s:%S)!", hp, pp));
```
Comment (JohnLZeller) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java]: `Created new DogStatsD client (%s:%S)!`

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java [hunk 6]
```diff
@@ -153,6 +158,31 @@ public final void onCompleted(final Run run, @Nonnull final TaskListener listene
       } else {
         serviceCheck("jenkins.job.status", DatadogBuildListener.CRITICAL, builddata, extraTags);
       }
+
+      //At this point. We will always have some tags. So no defensive checks needed.
```
Comment (JohnLZeller) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java]: nit: could we add a space before the words in comments? `// At this...`

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java [hunk 7]
```diff
@@ -397,6 +451,34 @@ public FormValidation doTestHostname(@QueryParameter("hostname") final String fo
...
+      }
+
+      //Match ports [1024-65535]
+      Pattern p = Pattern.compile("^(102[4-9]|10[3-9]\\d|1[1-9]\\d{2}|[2-9]\\d{3}|[1-5]\\d{4}|6[0-4]"
+              + "\\d{3}|65[0-4]\\d{2}|655[0-2]\\d|6553[0-5])$");
+      if(!p.matcher(pn).find()) {
+        return FormValidation.error("Invalid port specified. Range must be 1024-65535");
+      }
+
+      return FormValidation.ok("Everything ok");
```
Comment (JohnLZeller) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogBuildListener.java]: `Valid specification`

**Comment by JohnLZeller** [comment]
@MadsNielsen okay there is my first round of feedback. I'm gonna run some more tests on this :)

**Comment by MadsNielsen** [comment]
@JohnLZeller Great! I tried as best to test this, i think i had a setup with like 10 jobs running at intervals pushing events to datadog. I'll wait ammending this PR until you've finished your validation inhouse so that we can finish it up in a single go.

**Comment by JohnLZeller** [comment]
@MadsNielsen looks good, other than the nitpicks :)

**Comment by MadsNielsen** [comment]
@JohnLZeller great! Im on vacation now and Will be back after next weekend. I'll fix it then. Is this ok for your release schedule.?

**Comment by JohnLZeller** [comment]
@MadsNielsen we were hoping to release this a bit sooner actually. Would you mind if I rebased your commit and made the small changes? The commit will show up as a pair commit, but you'll still have attribution on the things you changed, and in the commit user.

**Comment by JohnLZeller** [comment]
@MadsNielsen here is the final PR, with your permission I'd like to merge this https://github.com/DataDog/jenkins-datadog-plugin/pull/43

**Comment by JohnLZeller** [comment]
Merged #43

**Comment by MadsNielsen** [comment]
Great you got it fixed! @JohnLZeller im perfectly fine with you fixing the issues you pointed out in your review while i am away😁

**Pull request merged**